File: src/fsharp/examples/curand/EstimatePi.fs
module Tutorial.Fs.examples.curand.EstimatePi

open System
open Alea.CUDA
open Alea.CUDA.CULib
open Alea.CUDA.Utilities.Timing
open NUnit.Framework
open CURANDInterop

// Target value
let PI = 3.14159265359

(*** define:cuRANDReduceSum ***)
let [<ReflectedDefinition>] reduceSum in' = 
    let sdata = __shared__.ExternArray<int>()

    // Perform first level of reduction:
    // - Write to shared memory
    let ltid = threadIdx.x

    sdata.[ltid] <- in'
    __syncthreads()

    // Do reduction in shared mem
    let mutable s = blockDim.x / 2
    while s > 0 do
        if ltid < s then sdata.[ltid] <- sdata.[ltid] + sdata.[ltid + s]
        __syncthreads()
        s <- s >>> 1
    
    sdata.[0]

(*** define:cuRANDComputeValue ***)
let computeValue =
    <@ fun (results:deviceptr<float>) (points:deviceptr<float>) (numSims:int) ->
        // Determine thread ID
        let bid = blockIdx.x
        let tid = blockIdx.x * blockDim.x + threadIdx.x
        let step = gridDim.x * blockDim.x

        // Shift the input/output pointers
        let mutable pointx = points + tid
        let mutable pointy = pointx + numSims
        
        // Count the number of points which lie inside the unit quarter-circle
        let mutable pointsInside = 0

        let mutable i = tid
        while i < numSims do
            let x = pointx.[0]
            let y = pointy.[0]
            let l2norm2 = x * x + y * y

            if l2norm2 < 1.0 then pointsInside <- pointsInside + 1

            i <- i + step; pointx <- pointx + step; pointy <- pointy + step

        // Reduce within the block
        pointsInside <- reduceSum pointsInside

        // Store the result
        if threadIdx.x = 0 then results.[bid] <- float pointsInside
    @>

(*** define:cuRANDPiEstimator ***)
let piEstimator (worker:Worker) numSims threadBlockSize =

    // Aim to launch around ten or more times as many blocks as there
    // are multiprocessors on the target device.
    let blocksPerSM = 10
    let numSMs = worker.Device.Attributes.MULTIPROCESSOR_COUNT

    // Determine how to divide the work between cores
    let block = dim3 threadBlockSize
    let grid = 
        let mutable x = (numSims + threadBlockSize - 1) / threadBlockSize
        while x > 2 * blocksPerSM * numSMs do x <- x >>> 1
        dim3(x)    

    let program = cuda {
        let! kernel = computeValue |> Compiler.DefineKernel
        
        return Entry(fun program ->
            let kernel = program.Apply kernel
            
            // Allocate memory for points
            // Each simulation has two random numbers to give X and Y coordinate
            let n = 2 * numSims
            let dPoints = worker.Malloc<float>(n)

            // Allocate memory for result
            // Each thread block will produce one result
            let dResults = worker.Malloc<float>(grid.x)

            // Generate random points in unit square
            let curand = new CURAND(worker, curandRngType_t.CURAND_RNG_QUASI_SOBOL64)
            curand.SetQuasiRandomGeneratorDimensions(2u)
            curand.SetGeneratorOrdering(curandOrdering_t.CURAND_ORDERING_QUASI_DEFAULT)
            curand.GenerateUniformDouble(dPoints.Ptr, size_t n)

            fun () ->
                let lp = LaunchParam(grid, block, block.x * sizeof<uint32>)
                kernel.Launch lp dResults.Ptr dPoints.Ptr numSims
                // Return partial results
                dResults.Gather()
            
                    )} |> worker.LoadProgram

    worker.Eval <| fun _ ->
        // Count the points inside unit quarter-circle
        let results = program.Run()

        // Complete sum-reduction on host
        let value = results |> Array.sum

        // Determine the propertion of points inside the quarter-circle,
        // i.e. the area of the unit quarter-circle
        let value = value / float numSims

        // Value is currently an estimate of the area of a unit quarter-circle, so we can
        // scale to a full circle by multiplying by four.  Now since the area of a circle
        // is pi * r^2, and r is one, the value will be an estimate for the value of pi.
        value * 4.0

(*** define:cuRANDEstimatePiTest ***)
let [<Test>] estimatePi() =
    let numSims = 100000
    let threadBlockSize = 128
    
    // Evaluate on GPU
    let worker = Worker.Default
    printfn "Estimating Pi on GPU (%s)\n" (worker.Device.Name)
    let estimator = piEstimator worker numSims threadBlockSize
    let result, t = tictoc(fun _ -> piEstimator worker numSims threadBlockSize)
    let elapsedTime = t.TotalMilliseconds

    // Tolerance to compare result with expected
    // This is just to check that nothing has gone very wrong with the
    // test, the actual accuracy of the result depends on the number of
    // Monte Carlo trials
    let tol = 0.01

    // Display results
    let abserror = abs(result - PI)
    let relerror = abserror / PI
    printfn "Precision:         %s" "double"
    printfn "Number of sims:    %d" numSims
    printfn "Tolerance:         %e" tol
    printfn "GPU result:        %e" result
    printfn "Expected:          %e" PI
    printfn "Absolute error:    %e" abserror
    printfn "Relative error:    %e" relerror

    // Check result
    if relerror > tol then printfn "computed result (%e) does not match expected result (%e)." result PI
    
    // Print results
    printfn "MonteCarloEstimatePiQ, Performance = %.2f sims/s, Time = %.2f(ms), NumDevsUsed = %u, Blocksize = %u"
        (float numSims / (elapsedTime/1000.0)) elapsedTime 1 threadBlockSize
    
    Assert.AreEqual(PI, result, tol)

File: src/csharp/examples/curand/EstimatePi.cs
using System;
using System.Diagnostics;
using System.Linq;
using Alea.CUDA;
using Alea.CUDA.CULib;
using Alea.CUDA.IL;
using NUnit.Framework;

namespace Tutorial.Cs.examples.curand
{
    
    internal class PiEstimatorModule : ILGPUModule
    {

        public PiEstimatorModule(GPUModuleTarget target)
            : base(target)
        {

        }

        //[cuRANDReduceSum]
        private int ReduceSum(int x)
        {
            var sdata = __shared__.ExternArray<int>();
            var ltid = threadIdx.x;
            sdata[ltid] = x;
            Intrinsic.__syncthreads();

            for (var s = blockDim.x / 2; s > 0; s >>= 1)
            {
                if (ltid < s) sdata[ltid] += sdata[ltid + s];
                Intrinsic.__syncthreads();
            }

            return sdata[0];
        }
        //[/cuRANDReduceSum]

        //[cuRANDComputeValue]
        [Kernel]
        public void ComputeValue(deviceptr<double> results, deviceptr<double> points, int numSims)
        {
            var bid = blockIdx.x;
            var tid = blockIdx.x * blockDim.x + threadIdx.x;
            var step = gridDim.x * blockDim.x;

            var pointx = points + tid;
            var pointy = pointx + numSims;

            var pointsInside = 0;

            for (var i = tid; i < numSims; i += step, pointx += step, pointy += step)
            {
                var x = pointx[0];
                var y = pointy[0];
                var l2norm2 = x * x + y * y;
                if (l2norm2 < 1.0) pointsInside++;
            }

            pointsInside = ReduceSum(pointsInside);

            if (threadIdx.x == 0) results[bid] = pointsInside;
        }
        //[/cuRANDComputeValue]

        //[cuRANDPiEstimator]
        public double RunEstimation(int numSims, int threadBlockSize)
        {
            const int blocksPerSm = 10;
            var numSMs = GPUWorker.Device.Attributes.MULTIPROCESSOR_COUNT;

            var block = new dim3(threadBlockSize);
            var grid = new dim3((numSims + threadBlockSize - 1) / threadBlockSize);
            while (grid.x > 2 * blocksPerSm * numSims) grid.x >>= 1;

            var n = 2 * numSims;
            using (var dPoints = GPUWorker.Malloc<double>(n))
            using (var dResults = GPUWorker.Malloc<double>(grid.x))
            {
                var curand = new CURAND(GPUWorker, CURANDInterop.curandRngType.CURAND_RNG_QUASI_SOBOL64);
                curand.SetQuasiRandomGeneratorDimensions(2);
                curand.SetGeneratorOrdering(CURANDInterop.curandOrdering.CURAND_ORDERING_QUASI_DEFAULT);
                curand.GenerateUniformDouble(dPoints.Ptr, new IntPtr(n));

                var lp = new LaunchParam(grid, block, block.x * sizeof(uint));
                GPULaunch(ComputeValue, lp, dResults.Ptr, dPoints.Ptr, numSims);
                var value = dResults.Gather().Sum();
                return (value/numSims)*4.0;
            }
        }
        //[/cuRANDPiEstimator]
    }

    [AOTCompile]
    class DefaultPiEstimatorModule : PiEstimatorModule
    {
        public DefaultPiEstimatorModule(GPUModuleTarget target)
            : base(target)
        {
        }

        private static DefaultPiEstimatorModule Instance;

        public static DefaultPiEstimatorModule DefaultInstance
        {
            get { return Instance ?? (Instance = new DefaultPiEstimatorModule(GPUModuleTarget.DefaultWorker)); }
        }
    }

    //[cuRANDEstimatePiTest]
    public class Test
    {
        // Target value
        private const double PI = 3.14159265359;

        [Test]
        public static void EstimatePi()
        {
            const int numSims = 100000;
            const int threadBlockSize = 128;
            var worker = GPUModuleTarget.DefaultWorker;
            Console.WriteLine("Estimating PI on GPU {0}\n", worker.GetWorker().Device.Name);
            var watch = Stopwatch.StartNew();
            var result = 
                DefaultPiEstimatorModule.DefaultInstance
                    .RunEstimation(numSims, threadBlockSize);
            watch.Stop();
            var elapsedTime = watch.Elapsed.TotalMilliseconds;
            const double tol = 0.01;
            var abserror = Math.Abs(result - Math.PI);
            var relerror = abserror / Math.PI;
            Console.WriteLine("Preceision:          {0}", "double");
            Console.WriteLine("Number of sims:      {0}", numSims);
            Console.WriteLine("Tolerance:           {0}", tol);
            Console.WriteLine("GPU result:          {0}", result);
            Console.WriteLine("Expected:            {0}", PI);
            Console.WriteLine("Absolute error:      {0}", abserror);
            Console.WriteLine("Relative error:      {0}", relerror);

            if (relerror > tol)
                Console.WriteLine("computed result ({0}) does not match expected result ({1}).", result, PI);

            Console.WriteLine("MonteCarloEstimatePiQ, Performance = {0} sims/s, Time = {1}(ms), NumDevsUsed = {2}, Blocksize = {3}",
                numSims / (elapsedTime/1000.0), elapsedTime, 1, threadBlockSize);

            Assert.AreEqual(PI, result, tol);
        }
        //[/cuRANDEstimatePiTest]
    }
}
